File: cachelru/src/main/java/uv/lis/CacheLRU.java
package uv.lis;

import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.Deque;
import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

/**
 * Cache generica con politica de reemplazo LRU (Least Recently Used).
 *
 * <p>Integra dos estructuras de datos de la biblioteca estandar de Java:</p>
 * <ul>
 *   <li>{@link HashMap} (tabla hash): mapea clave -&gt; valor en O(1) promedio.</li>
 *   <li>{@link Deque} implementado con {@link LinkedList} (cola doblemente
 *       enlazada): mantiene el orden de uso. El primer elemento es el menos
 *       recientemente usado (LRU) y el ultimo es el mas reciente (MRU).</li>
 * </ul>
 *
 * <p>La combinacion permite que {@code get} y {@code put} operen en O(1)
 * promedio para la busqueda y la expulsion; reubicar una clave usada al
 * extremo MRU recorre la cola, por lo que es O(n).</p>
 *
 * @param <K> tipo de las claves.
 * @param <V> tipo de los valores.
 */
public class CacheLRU<K, V> {

    private final int capacidad;
    private final Map<K, V> tabla;     // tabla hash: clave -> valor
    private final Deque<K> orden;      // orden de uso: primero = LRU, ultimo = MRU

    private int aciertos;
    private int fallos;
    private int expulsiones;

    private final RegistroOperaciones registro;

    /**
     * Crea una cache sin registro de operaciones.
     *
     * @param capacidad capacidad maxima (entero positivo).
     */
    public CacheLRU(int capacidad) {
        this(capacidad, null);
    }

    /**
     * Crea una cache con un registro de operaciones opcional.
     *
     * @param capacidad capacidad maxima (entero positivo).
     * @param registro  registro de operaciones, o {@code null} para no registrar.
     * @throws IllegalArgumentException si la capacidad no es un entero positivo.
     */
    public CacheLRU(int capacidad, RegistroOperaciones registro) {
        if (capacidad <= 0) {
            throw new IllegalArgumentException("La capacidad debe ser un entero positivo.");
        }
        this.capacidad = capacidad;
        this.tabla = new HashMap<>();
        this.orden = new LinkedList<>();
        this.registro = registro;
        this.aciertos = 0;
        this.fallos = 0;
        this.expulsiones = 0;
    }

    // ------------------------------------------------------------------ //
    // Consultas de estado                                                //
    // ------------------------------------------------------------------ //

    public int tamano() { return tabla.size(); }

    public int getCapacidad() { return capacidad; }

    public boolean estaLlena() { return tabla.size() >= capacidad; }

    public boolean estaVacia() { return tabla.isEmpty(); }

    public int getAciertos() { return aciertos; }

    public int getFallos() { return fallos; }

    public int getExpulsiones() { return expulsiones; }

    // ------------------------------------------------------------------ //
    // Apoyo interno                                                      //
    // ------------------------------------------------------------------ //

    /**
     * Mueve una clave existente al extremo MRU (final) de la cola.
     *
     * <p>La cola no permite acceso directo a un elemento interno, por lo que
     * {@code remove} la recorre: esta operacion es O(n). Es el costo de usar
     * una cola nativa en lugar de una lista enlazada con referencias directas
     * a los nodos.</p>
     */
    private void marcarComoReciente(K clave) {
        orden.remove(clave);
        orden.addLast(clave);
    }

    private void anotar(String operacion, K clave, String resultado, String detalle) {
        if (registro != null) {
            registro.registrar(operacion, String.valueOf(clave), resultado, detalle);
        }
    }

    // ------------------------------------------------------------------ //
    // Operaciones fundamentales                                          //
    // ------------------------------------------------------------------ //

    /**
     * Consulta el valor asociado a una clave.
     *
     * <p>En caso de acierto marca el elemento como el mas recientemente usado y
     * devuelve su valor; en caso de fallo devuelve {@code null}.</p>
     *
     * @param clave clave a consultar.
     * @return el valor asociado, o {@code null} si no esta en la cache.
     */
    public V get(K clave) {
        if (tabla.containsKey(clave)) {
            marcarComoReciente(clave);
            aciertos++;
            V valor = tabla.get(clave);
            anotar("GET", clave, "ACIERTO", "valor=" + valor);
            return valor;
        }
        fallos++;
        anotar("GET", clave, "FALLO", "clave no encontrada");
        return null;
    }

    /**
     * Inserta o actualiza un par clave-valor.
     *
     * <p>Si la clave ya existe, actualiza su valor y la marca como MRU. Si es
     * nueva y la cache esta llena, expulsa primero el elemento LRU.</p>
     *
     * @param clave clave a insertar o actualizar.
     * @param valor valor asociado.
     */
    public void put(K clave, V valor) {
        // Caso 1: la clave ya existe -> actualizar, sin expulsar.
        if (tabla.containsKey(clave)) {
            tabla.put(clave, valor);
            marcarComoReciente(clave);
            anotar("PUT", clave, "ACTUALIZA", "valor=" + valor);
            return;
        }
        // Caso 2: clave nueva y cache llena -> expulsar el LRU (caso limite).
        if (estaLlena()) {
            K lru = orden.removeFirst();   // O(1): primer elemento = LRU
            tabla.remove(lru);
            expulsiones++;
            anotar("PUT", clave, "EXPULSA", "se elimino LRU clave=" + lru);
        }
        // Caso 3: insertar el nuevo par y marcarlo como MRU.
        tabla.put(clave, valor);
        orden.addLast(clave);
        anotar("PUT", clave, "INSERTA", "valor=" + valor);
    }

    // ------------------------------------------------------------------ //
    // Reportes                                                           //
    // ------------------------------------------------------------------ //

    /**
     * Devuelve el contenido de la cache ordenado del mas reciente (MRU) al
     * menos reciente (LRU).
     *
     * @return lista de pares (clave, valor).
     */
    public List<Map.Entry<K, V>> contenido() {
        List<Map.Entry<K, V>> lista = new ArrayList<>();
        Iterator<K> it = orden.descendingIterator();   // del MRU (ultimo) al LRU (primero)
        while (it.hasNext()) {
            K clave = it.next();
            lista.add(new AbstractMap.SimpleEntry<>(clave, tabla.get(clave)));
        }
        return lista;
    }

    /**
     * Devuelve un resumen inmutable de las metricas acumuladas.
     *
     * @return objeto {@link Estadisticas}.
     */
    public Estadisticas estadisticas() {
        return new Estadisticas(capacidad, tabla.size(), aciertos, fallos, expulsiones);
    }
}

File: cache_lru.py
"""
cache_lru.py
============
Cache con politica de reemplazo LRU (Least Recently Used) construida sobre dos
estructuras de datos de la biblioteca estandar de Python:

    * dict              -> tabla hash: mapea clave -> valor en O(1) promedio.
    * collections.deque -> cola doblemente enlazada: mantiene el orden de uso.

Se emplean las estructuras nativas (en lugar de implementarlas desde cero)
porque el problema corresponde al dominio de Sistemas Operativos y el uso
correcto de las estructuras es suficiente. Internamente, 'dict' es una tabla
hash y 'deque' es una lista doblemente enlazada, por lo que se corresponden con
las estructuras descritas en el marco teorico (tabla hash + cola para LRU).

Convencion de orden en la deque:
    * Extremo izquierdo -> elemento MENOS recientemente usado (LRU).
    * Extremo derecho   -> elemento MAS recientemente usado (MRU).
"""

import logging
from collections import deque


class CacheLRU:
    """Cache de capacidad fija con politica de reemplazo LRU."""

    def __init__(self, capacidad, registrar=True):
        """Crea la cache.

        Parametros
        ----------
        capacidad : int
            Numero maximo de elementos. Debe ser un entero positivo.
        registrar : bool
            Si es True, cada operacion se registra mediante el modulo logging.

        Lanza ValueError si la capacidad no es un entero positivo (caso limite).
        """
        if not isinstance(capacidad, int) or isinstance(capacidad, bool) or capacidad <= 0:
            raise ValueError("La capacidad debe ser un entero positivo.")

        self.capacidad = capacidad
        self._tabla = {}            # tabla hash: clave -> valor
        self._orden = deque()       # orden de uso: izquierda = LRU, derecha = MRU

        # Metricas para los reportes de operaciones.
        self.aciertos = 0
        self.fallos = 0
        self.expulsiones = 0

        self._registrar = registrar
        self._log = logging.getLogger("cache_lru")

    # ------------------------------------------------------------------ #
    # Consultas de estado                                                #
    # ------------------------------------------------------------------ #
    def __len__(self):
        return len(self._tabla)

    def esta_llena(self):
        return len(self) >= self.capacidad

    def esta_vacia(self):
        return len(self) == 0

    # ------------------------------------------------------------------ #
    # Apoyo interno                                                      #
    # ------------------------------------------------------------------ #
    def _marcar_como_reciente(self, clave):
        """Mueve una clave existente al extremo MRU (derecha) de la deque.

        La deque no permite acceso directo a un elemento interno, por lo que
        remove() recorre la cola: esta operacion es O(n). Es el costo de usar
        una cola nativa en lugar de una lista enlazada con referencias directas
        a los nodos (que lograria O(1)). Para el tamano tipico de una cache es
        despreciable.
        """
        self._orden.remove(clave)
        self._orden.append(clave)

    # ------------------------------------------------------------------ #
    # Operaciones fundamentales                                          #
    # ------------------------------------------------------------------ #
    def get(self, clave):
        """Consulta el valor de una clave.

        En caso de ACIERTO marca el elemento como el mas recientemente usado y
        devuelve su valor. En caso de FALLO devuelve None.
        """
        if clave in self._tabla:
            self._marcar_como_reciente(clave)
            self.aciertos += 1
            valor = self._tabla[clave]
            self._anotar("GET", clave, "ACIERTO", "valor=%s" % valor)
            return valor

        self.fallos += 1
        self._anotar("GET", clave, "FALLO", "clave no encontrada")
        return None

    def put(self, clave, valor):
        """Inserta o actualiza un par clave-valor.

        Si la clave ya existe, actualiza su valor y la marca como MRU.
        Si es nueva y la cache esta llena, expulsa primero el elemento LRU.
        """
        # Caso 1: la clave ya existe -> actualizar, sin expulsar.
        if clave in self._tabla:
            self._tabla[clave] = valor
            self._marcar_como_reciente(clave)
            self._anotar("PUT", clave, "ACTUALIZA", "valor=%s" % valor)
            return

        # Caso 2: clave nueva y cache llena -> expulsar el LRU (caso limite).
        if self.esta_llena():
            lru = self._orden.popleft()         # O(1): extremo izquierdo = LRU
            del self._tabla[lru]
            self.expulsiones += 1
            self._anotar("PUT", clave, "EXPULSA", "se elimino LRU clave=%s" % lru)

        # Caso 3: insertar el nuevo par y marcarlo como MRU.
        self._tabla[clave] = valor
        self._orden.append(clave)
        self._anotar("PUT", clave, "INSERTA", "valor=%s" % valor)

    # ------------------------------------------------------------------ #
    # Reportes                                                           #
    # ------------------------------------------------------------------ #
    def contenido(self):
        """Devuelve la lista de pares (clave, valor) ordenada de MRU a LRU."""
        return [(clave, self._tabla[clave]) for clave in reversed(self._orden)]

    def estadisticas(self):
        """Devuelve un diccionario con las metricas acumuladas."""
        total = self.aciertos + self.fallos
        tasa = (self.aciertos / total * 100) if total else 0.0
        return {
            "capacidad": self.capacidad,
            "elementos": len(self),
            "aciertos": self.aciertos,
            "fallos": self.fallos,
            "expulsiones": self.expulsiones,
            "consultas_totales": total,
            "tasa_aciertos_%": round(tasa, 2),
        }

    def _anotar(self, operacion, clave, resultado, detalle=""):
        """Registra una operacion en el log si el registro esta activado."""
        if self._registrar:
            self._log.info("%-4s clave=%-8s -> %-9s %s",
                           operacion, str(clave), resultado, detalle)